### PR TITLE
implemented notification options store

### PR DESCRIPTION
**Pull request by DFSko**
implemented agent notification options resolves happy-travel/agent-app-project#56

### CODE REVIEW

File: Api/Models/Notifications/NotificationOptionSlim.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+using System.Collections.Generic;
+using HappyTravel.Edo.Common.Enums.Notifications;
+
+namespace HappyTravel.Edo.Api.Models.Notifications
+{
+    public readonly struct NotificationOptionSlim
```
Comment (mingazhev) [inline Api/Models/Notifications/NotificationOptionSlim.cs:6]: SlimNotificationOptions

File: HappyTravel.Edo.Common/Enums/Notifications/NotificationType.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
+namespace HappyTravel.Edo.Common.Enums.Notifications
+{
+    public enum NotificationType
+    {
+        BookingVoucher,
```
Comment (mingazhev) [inline HappyTravel.Edo.Common/Enums/Notifications/NotificationType.cs:5]: Enum numbers

File: Api/Models/Notifications/NotificationOptionSlim.cs (hunk 1 above)
Comment (DFSko) [reply Api/Models/Notifications/NotificationOptionSlim.cs]: Fixed

File: HappyTravel.Edo.Common/Enums/Notifications/NotificationType.cs (hunk 2 above)
Comment (DFSko) [reply HappyTravel.Edo.Common/Enums/Notifications/NotificationType.cs]: Fixed

File: Api/Services/Notifications/NotificationOptionsService.cs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    {
+        public NotificationOptionsService(EdoContext context)
+        {
+            _context = context;
+        }
+
+
+        public async Task<Result<SlimNotificationOption>> GetNotificationOptions(NotificationType type, AgentContext agent)
+        {
+            var option = await GetOption(type, agent.AgentId, agent.AgencyId);
```
Comment (mingazhev) [inline Api/Services/Notifications/NotificationOptionsService.cs:23]: GetOptions, this is not a single option

File: Api/Services/Notifications/NotificationOptionsService.cs [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        }
+
+
+        public Task<Result> Update(NotificationType type, SlimNotificationOption option, AgentContext agentContext)
+        {
+            return CheckMandatory()
+                .Bind(SaveOption);
+
+
+            Result<SlimNotificationOption> CheckMandatory()
```
Comment (mingazhev) [inline Api/Services/Notifications/NotificationOptionsService.cs:37]: I think just Validate will be enough. CheckMandatory does not sound clear for me

File: Api/Services/Notifications/NotificationOptionsService.cs [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+            }
+
+
+            async Task<Result> SaveOption(SlimNotificationOption defaultOption)
+            {
+                var entity = await GetOption(type, agentContext.AgentId, agentContext.AgencyId);
+
+                if (entity is null)
+                {
+                    _context.NotificationOptions.Add(new NotificationOption
```
Comment (mingazhev) [inline Api/Services/Notifications/NotificationOptionsService.cs:56]: Object that contains more than one field as settings or option should be named as plural

File: HappyTravel.Edo.Common/Enums/Notifications/NotificationType.cs [hunk 6]
```diff
@@ -0,0 +1,15 @@
+namespace HappyTravel.Edo.Common.Enums.Notifications
+{
+    public enum NotificationType
+    {
+        None = 0,
+        BookingVoucher = 1,
+        BookingInvoice = 2,
+        DeadlineApproaching = 3,
+        SuccessfulPaymentReceipt = 4,
+        BookingDuePayment = 5,
```
Comment (mingazhev) [inline HappyTravel.Edo.Common/Enums/Notifications/NotificationType.cs:10]: looks like Date postfix is missing there

File: Api/Services/Notifications/NotificationOptionsService.cs [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+
+        private Task<NotificationOption> GetOption(NotificationType type, int agentId, int agencyId) 
+            => _context.NotificationOptions
+                .SingleOrDefaultAsync(o => o.AgencyId == agencyId && o.AgentId == agentId && o.Type == type);
+
+
+        private readonly Dictionary<NotificationType, SlimNotificationOption> _defaultOptions = new()
+        {
+            {NotificationType.BookingVoucher, new() {EnabledProtocols = ProtocolTypes.Email | ProtocolTypes.WebSocket, IsMandatory = true}},
```
Comment (kirillta) [inline Api/Services/Notifications/NotificationOptionsService.cs:89]: What if only one type is mandatory?

File: Api/Services/Notifications/NotificationOptionsService.cs (hunk 3 above)
Comment (DFSko) [reply Api/Services/Notifications/NotificationOptionsService.cs]: Fixed

File: Api/Services/Notifications/NotificationOptionsService.cs (hunk 4 above)
Comment (DFSko) [reply Api/Services/Notifications/NotificationOptionsService.cs]: Fixed

File: Api/Services/Notifications/NotificationOptionsService.cs (hunk 5 above)
Comment (DFSko) [reply Api/Services/Notifications/NotificationOptionsService.cs]: Fixed

File: HappyTravel.Edo.Common/Enums/Notifications/NotificationType.cs (hunk 6 above)
Comment (DFSko) [reply HappyTravel.Edo.Common/Enums/Notifications/NotificationType.cs]: Fixed

File: HappyTravel.Edo.Common/Enums/Notifications/NotificationType.cs [hunk 8]
```diff
@@ -0,0 +1,15 @@
+namespace HappyTravel.Edo.Common.Enums.Notifications
+{
+    public enum NotificationType
```
[the remark this answers is not in the document]
Comment (DFSko) [reply HappyTravel.Edo.Common/Enums/Notifications/NotificationType.cs]: Fixed

File: Api/Services/Notifications/NotificationOptionsService.cs (hunk 7 above)
Comment (DFSko) [reply Api/Services/Notifications/NotificationOptionsService.cs]: In validation method checks that at least one flag is set

**Review by mingazhev — APPROVED** [review]

**Pull request merged**